libraries/networking/src/AssetClient.cpp

```diff
@@ -0,0 +1,208 @@
+//
+//  AssetClient.cpp
+//
+//  Created by Ryan Huffman on 2015/07/21
+//  Copyright 2015 High Fidelity, Inc.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+#include "AssetClient.h"
+
+#include <QBuffer>
+#include <QThread>
+
+#include "AssetRequest.h"
+#include "NodeList.h"
+#include "PacketReceiver.h"
+
+MessageID AssetClient::_currentID = 0;
+
+
+AssetClient::AssetClient() {
+    auto& packetReceiver = DependencyManager::get<NodeList>()->getPacketReceiver();
+    packetReceiver.registerListener(PacketType::AssetGetInfoReply, this, "handleAssetGetInfoReply");
+    packetReceiver.registerMessageListener(PacketType::AssetGetReply, this, "handleAssetGetReply");
+    packetReceiver.registerListener(PacketType::AssetUploadReply, this, "handleAssetUploadReply");
+}
+
+AssetRequest* AssetClient::create(QString hash) {
+    if (QThread::currentThread() != thread()) {
+        AssetRequest* req;
+        QMetaObject::invokeMethod(this, "create",
+            Qt::BlockingQueuedConnection, 
+            Q_RETURN_ARG(AssetRequest*, req),
+            Q_ARG(QString, hash));
+        return req;
+    }
+
+    if (hash.length() != 32) {
+        qDebug() << "Invalid hash size";
+        return nullptr;
+    }
+
+    auto nodeList = DependencyManager::get<NodeList>();
+    SharedNodePointer assetServer = nodeList->soloNodeOfType(NodeType::AssetServer);
+
+    if (assetServer) {
+        auto assetClient = DependencyManager::get<AssetClient>();
+        auto request = new AssetRequest(assetClient.data(), hash);
+
+        return request;
+    }
+
+    return nullptr;
+}
```

Comment:
any chance we should have this return an std::unique_ptr so the ownership transfer is clear?